### ISSUE TITLE
Issues with ImageRecordReaders and Multiple Subfolders

### ISSUE THREAD

**Issue by ChrisHayduk** [deeplearning4j/deeplearning4j#1507]
So right now I am trying to read in 18 different classes with 60 images per class. These files are contained in subfolders (with each subfolder being the name of the class), with all of the subfolders contained in a main "training" folder. When using my neural network and ImageRecordReader, I keep having only one class predicted, no matter how I tune the net or what data I use. I realized that the file structure may be the issue, with ImageRecorReader not being able to properly recognize it or deal with it. 

I set up the method to read in my images using the Image Data Pipeline tutorial page:
[http://deeplearning4j.org/image-data-pipeline](http://deeplearning4j.org/image-data-pipeline)

When ImageRecordReader is attempting to read in the images, I think there may be one of two issues - it is either not reading in the images correctly (due to not accessing each individual subfolder) or it is not assigning the class labels to each image.

The link to a gist and my attached training data are below:

[Gist](https://gist.github.com/ChrisHayduk/90063e636d9831e2532b10f861bce169)

[training.zip](https://github.com/deeplearning4j/deeplearning4j/files/251028/training.zip)

**Comment by WikiDreams** [comment]
Here is another example with the same problem I think.
`public class Cifar {

	/* Images folder structure
	 * img/frog
	 * img/dog
	 * ...
	 * 10 classes each with 5000 images (32X32).
	 */
	
	
	private static final Logger log = LoggerFactory.getLogger(Cifar.class);

	private static File trainImagesFolder = new File("C:/Development/DeepLearning/DataSets/CIFAR-10/img");
	
	private static final int WIDTH = 32;
	private static final int HEIGHT = 32;
	
	private static int ITERATIONS = 1;
	private static int listenerFreq = ITERATIONS / 5;
	private static int SEED = 123;
	private static int BATCHSIZE = 1000;
	private static int labelIndex = WIDTH * HEIGHT;

	private static ArrayList<String> LABELS = new ArrayList<>();
	
	private static int splitTrainNum = (int) (BATCHSIZE * 0.8);
	private static SplitTestAndTrain trainAndTest;
	private static DataSet trainInput;
	private static List<INDArray> testInput = new ArrayList<>();

	public static void run() throws Exception {

		log.info("Load images");
		
		for (File f : trainImagesFolder.listFiles()) {
		LABELS.add(f.getName());	
		}
		
		RecordReader imageReader = new ImageRecordReader(WIDTH, HEIGHT, true, LABELS);
		imageReader.initialize(new FileSplit(trainImagesFolder));

		DataSetIterator dataSetIterator = new RecordReaderDataSetIterator(imageReader, BATCHSIZE, labelIndex, LABELS.size());

		log.info("Get network configuration.");
		MultiLayerNetwork model = new MultiLayerNetwork(getConfiguration());
		model.init();
		model.setListeners(Arrays.asList((IterationListener) new ScoreIterationListener(listenerFreq)));

		log.info("Train model");
		DataSet cifarDataSet;
		while (dataSetIterator.hasNext()) {
			cifarDataSet = dataSetIterator.next();
trainAndTest = cifarDataSet.splitTestAndTrain(splitTrainNum, new Random(SEED));
trainInput = trainAndTest.getTrain();
testInput.add(trainAndTest.getTest().getFeatureMatrix());
			model.fit(trainInput);
		}

		log.info("Evaluate model");
		Evaluation eval = new Evaluation(LABELS.size());
		for (int i = 0; i < testInput.size(); i++) {
			INDArray output = model.output(testInput.get(i));
		}

		log.info(eval.stats());
	}


	public static MultiLayerConfiguration getConfiguration() {
		MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
				.seed(SEED)
				.iterations(ITERATIONS)
				.optimizationAlgo(OptimizationAlgorithm.LINE_GRADIENT_DESCENT)
				.list(10)
				.layer(0, new RBM.Builder().nIn(WIDTH * HEIGHT).nOut(1000).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(1, new RBM.Builder().nIn(1000).nOut(500).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(2, new RBM.Builder().nIn(500).nOut(250).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(3, new RBM.Builder().nIn(250).nOut(100).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(4, new RBM.Builder().nIn(100).nOut(30).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build()) //encoding stops
				.layer(5, new RBM.Builder().nIn(30).nOut(100).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build()) //decoding starts
				.layer(6, new RBM.Builder().nIn(100).nOut(250).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(7, new RBM.Builder().nIn(250).nOut(500).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(8, new RBM.Builder().nIn(500).nOut(1000).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
				.layer(9, new OutputLayer.Builder(LossFunctions.LossFunction.RMSE_XENT).nIn(1000).nOut(WIDTH * HEIGHT).build())
				.pretrain(true).backprop(true)
				.build();

		return conf;
	}

}
`

**Comment by WikiDreams** [comment]
[Here is another project with the same problem.](https://github.com/WikiDreams/DeepLearningImageRecorderIssue)

**Comment by saudet** [comment]
FYI, this might be the kind of thing I'll be looking into in priority over the next few days for https://github.com/deeplearning4j/Canova/issues/135...

**Comment by WikiDreams** [comment]
I'm anewbie using DL4J but I can help if something needed.

**Comment by ChrisHayduk** [comment]
@saudet Awesome! Let me know if there is anything I can do to help.

Also, do you know when this issue might be resolved? Or if there is any workaround I could use at the moment to get it working? I'm trying to submit a research paper to a conference by May 13th, so any if I could get this setup soon, that would be great.

**Comment by WikiDreams** [comment]
Do you already test with my class?

**Comment by ChrisHayduk** [comment]
@WikiDreams One of my other team members was testing your code. I'll ask him to see how it went.

**Comment by WikiDreams** [comment]
If you need, call me ;)

**Comment by saudet** [comment]
@ChrisHayduk This might help for your immediate needs: https://github.com/deeplearning4j/nd4j/issues/897#issuecomment-217495023

**Comment by ChrisHayduk** [comment]
@saudet Thanks for the link! I tried setting it up that way, and the accuracy for my neural net is still pretty terrible. It is only predicting 6 or 7 out of the 18 classes every time I run it, even though I am testing on the training data, so it shouldn't be too difficult to get 100% accuracy. I am not sure if this is an issue with my neural net itself, or if the labels are still being added incorrectly.

**Comment by agibsonccc** [comment]
@ChrisHayduk Have you tried repartitioning your dataset to ensure minibatches are even? I'm still not sure if there was ever a bug in the first place. I keep telling people you have to TUNE your neural nets. Your neural net failing isn't always a software bug.

**Comment by ajohnston9** [comment]
@agibsoncc From what I understand about my and @ChrisHayduk's situation,
the problem seems to be that ImageRecordReader's default behavior does
iterate over subfolder and pick up the images (or, at very most, only for
the last class). @ChrisHayduk's current problem, however, could be caused
by a variety of things.

**Comment by ChrisHayduk** [comment]
@agibsonccc I've been trying to tune my neural net for the past month, but none of the changes to any of the parameters seemed to affect the situation. No matter what I did, the net would only predict one class and sometimes would only assign one class label to all of the data. The change to the ImageRecordReader call (that saudet linked) is the only thing that has produced different results, so it is leading me to believe that ImageRecordReader is the issue.

**Comment by agibsonccc** [comment]
Have you actually been checking the batches going into the neural net
though? It's not like it's hard to so a while iterator .has next and just
print the labels going into it.

**Comment by ChrisHayduk** [comment]
@agibsonccc When I try to print next.getLabelNames(), it's just outputting empty brackets.

**Comment by agibsonccc** [comment]
Look at:
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/main/java/org/deeplearning4j/datasets/rearrange/LocalUnstructuredDataFormatter.java

If you want to do this in memory (load it all at once and then save it to local disk:
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java#L828

Don't use getLabelNames() you have to use:
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java#L389

**Comment by agibsonccc** [comment]
Give me a github gist with a print out of the labels and the like and show me what it's outputting. If the labels aren't even, use the methods above I showed you and see what it does. If the label output is the same as your local file structure then we've done our job. I'm just hesitant to call this a bug when the problem is likely something else..this is such an easy thing to have and we have a decent number of unit tests for this.

**Comment by ChrisHayduk** [comment]
@agibsonccc I'm sorry for the confusion. It seems like the labels are being output appropriately using the method that @saudet linked me to.

**Comment by WikiDreams** [comment]
Chris, what you when you change the connection parameters and the output was always the same? I've got this output for a total of 10000 images, 2 classes each one with 5000 images.
Examples labeled as 0 classified by model as 0: 5000 times
Examples labeled as 1 classified by model as 0: 5000 times

Warning: class 1 was never predicted by the model. This class was excluded from the average precision

==========================Scores========================================
 Accuracy:  0.5
 Precision: 0.5
 Recall:    0.5
 F1 Score:  0.5
If I change the parameters of the model the output is always the same... someone to help?
